TASK FOR WEEK 1 -  Include concepts such as linear regression with polynomial terms, interaction terms, multicollinearity, variance inflation factor and regression, and categorical and continuous features.

In [ ]:
import pandas as pd

In [ ]:
phishing_df_1 = pd.read_csv('/content/dataset_phishing.csv')
phishing_df_1.head()

url  length_url  \
0              http://www.crestonwood.com/router.php          37   
1  http://shadetreetechnology.com/V4/validation/a...          77   
2  https://support-appleld.com.secureupdate.duila...         126   
3                                 http://rgipt.ac.in          18   
4  http://www.iracing.com/tracks/gateway-motorspo...          55   

   length_hostname  ip  nb_dots  nb_hyphens  nb_at  nb_qm  nb_and  nb_or  ...  \
0               19   0        3           0      0      0       0      0  ...   
1               23   1        1           0      0      0       0      0  ...   
2               50   1        4           1      0      1       2      0  ...   
3               11   0        2           0      0      0       0      0  ...   
4               15   0        2           2      0      0       0      0  ...   

   domain_in_title  domain_with_copyright  whois_registered_domain  \
0                0                      1                        0   
1                1                      0                        0   
2                1                      0                        0   
3                1                      0                        0   
4                0                      1                        0   

   domain_registration_length  domain_age  web_traffic  dns_record  \
0                          45          -1            0           1   
1                          77        5767            0           0   
2                          14        4004      5828815           0   
3                          62          -1       107721           0   
4                         224        8175         8725           0   

   google_index  page_rank      status  
0             1          4  legitimate  
1             1          2    phishing  
2             1          0    phishing  
3             0          3  legitimate  
4             0          6  legitimate  

[5 rows x 89 columns]

In [ ]:
phishing_df_1.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11430 entries, 0 to 11429
Data columns (total 89 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   url                         11430 non-null  object 
 1   length_url                  11430 non-null  int64  
 2   length_hostname             11430 non-null  int64  
 3   ip                          11430 non-null  int64  
 4   nb_dots                     11430 non-null  int64  
 5   nb_hyphens                  11430 non-null  int64  
 6   nb_at                       11430 non-null  int64  
 7   nb_qm                       11430 non-null  int64  
 8   nb_and                      11430 non-null  int64  
 9   nb_or                       11430 non-null  int64  
 10  nb_eq                       11430 non-null  int64  
 11  nb_underscore               11430 non-null  int64  
 12  nb_tilde                    11430 non-null  int64  
 13  nb_percent                  114

In [ ]:
phishing_df_1.describe()


length_url  length_hostname            ip       nb_dots  \
count  11430.000000     11430.000000  11430.000000  11430.000000   
mean      61.126684        21.090289      0.150569      2.480752   
std       55.297318        10.777171      0.357644      1.369686   
min       12.000000         4.000000      0.000000      1.000000   
25%       33.000000        15.000000      0.000000      2.000000   
50%       47.000000        19.000000      0.000000      2.000000   
75%       71.000000        24.000000      0.000000      3.000000   
max     1641.000000       214.000000      1.000000     24.000000   

         nb_hyphens         nb_at         nb_qm        nb_and    nb_or  \
count  11430.000000  11430.000000  11430.000000  11430.000000  11430.0   
mean       0.997550      0.022222      0.141207      0.162292      0.0   
std        2.087087      0.155500      0.364456      0.821337      0.0   
min        0.000000      0.000000      0.000000      0.000000      0.0   
25%        0.000000      0.000000      0.000000      0.000000      0.0   
50%        0.000000      0.000000      0.000000      0.000000      0.0   
75%        1.000000      0.000000      0.000000      0.000000      0.0   
max       43.000000      4.000000      3.000000     19.000000      0.0   

              nb_eq  ...   empty_title  domain_in_title  \
count  11430.000000  ...  11430.000000     11430.000000   
mean       0.293176  ...      0.124759         0.775853   
std        0.998317  ...      0.330460         0.417038   
min        0.000000  ...      0.000000         0.000000   
25%        0.000000  ...      0.000000         1.000000   
50%        0.000000  ...      0.000000         1.000000   
75%        0.000000  ...      0.000000         1.000000   
max       19.000000  ...      1.000000         1.000000   

       domain_with_copyright  whois_registered_domain  \
count           11430.000000             11430.000000   
mean                0.439545                 0.072878   
std                 0.496353                 0.259948   
min                 0.000000                 0.000000   
25%                 0.000000                 0.000000   
50%                 0.000000                 0.000000   
75%                 1.000000                 0.000000   
max                 1.000000                 1.000000   

       domain_registration_length    domain_age   web_traffic    dns_record  \
count                11430.000000  11430.000000  1.143000e+04  11430.000000   
mean                   492.532196   4062.543745  8.567566e+05      0.020122   
std                    814.769415   3107.784600  1.995606e+06      0.140425   
min                     -1.000000    -12.000000  0.000000e+00      0.000000   
25%                     84.000000    972.250000  0.000000e+00      0.000000   
50%                    242.000000   3993.000000  1.651000e+03      0.000000   
75%                    449.000000   7026.750000  3.738455e+05      0.000000   
max                  29829.000000  12874.000000  1.076799e+07      1.000000   

       google_index     page_rank  
count  11430.000000  11430.000000  
mean       0.533946      3.185739  
std        0.498868      2.536955  
min        0.000000      0.000000  
25%        0.000000      1.000000  
50%        1.000000      3.000000  
75%        1.000000      5.000000  
max        1.000000     10.000000  

[8 rows x 87 columns]

In [ ]:
phishing_df_1.isnull().sum()

url                0
length_url         0
length_hostname    0
ip                 0
nb_dots            0
                  ..
web_traffic        0
dns_record         0
google_index       0
page_rank          0
status             0
Length: 89, dtype: int64

In [ ]:
X = phishing_df_1.drop('status', axis=1)
y = phishing_df_1['status']

In [ ]:
# Check data types of columns in X
print("Data types of X before VIF calculation:")
print(X.dtypes)

Data types of X before VIF calculation:
url                object
length_url          int64
length_hostname     int64
ip                  int64
nb_dots             int64
                    ...  
domain_age          int64
web_traffic         int64
dns_record          int64
google_index        int64
page_rank           int64
Length: 88, dtype: object


In [ ]:
import numpy as np

In [ ]:
X_numeric = X.select_dtypes(include=np.number)

In [ ]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.tools.tools import add_constant

In [ ]:
X_with_const = add_constant(X_numeric) # Use the numeric DataFrame. This adds a column of 1s (called "const") to your data so the model includes an intercept term.

vif = pd.DataFrame()
vif["Variable"] = X_with_const.columns #This part calculates the VIF for each variable (column).

# Calculate VIF, ensuring we use the numeric values
vif["VIF"] = [variance_inflation_factor(X_with_const.values, i) for i in range(X_with_const.shape[1])] #VIF shows how much a variable is related to other variables.
vif

/usr/local/lib/python3.11/dist-packages/statsmodels/regression/linear_model.py:1782: RuntimeWarning: invalid value encountered in scalar divide
  return 1 - self.ssr/self.centered_tss


Variable         VIF
0             const  124.747685
1        length_url   11.469510
2   length_hostname    3.941609
3                ip    3.413926
4           nb_dots    8.152036
..              ...         ...
83       domain_age    1.875603
84      web_traffic    1.244188
85       dns_record    1.652678
86     google_index    1.967224
87        page_rank    2.981246

[88 rows x 2 columns]

VIF > 10 means strong multicollinearity. That variable might be too similar to others and could mess up your model.

Example: length_url has a VIF of 11.47, so it might be giving similar information as other variables. You might consider removing or combining it with related features.

High VIF means the model might not be trustworthy.

Multicollinearity can make it hard to tell which variable is really important.



In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score


# Convert the target variable 'y' to numerical values
# Map 'phishing' to 1 and 'legitimate' to 0 (or vice versa)
y_encoded = y.map({'phishing': 1, 'legitimate': 0})

# Select only numeric columns for X
# Keep this as before, ensuring only numeric features are used for the model
X_numeric = X.select_dtypes(include=np.number)

# Split data using the numeric DataFrame for X and the encoded target variable y
X_train, X_test, y_train, y_test = train_test_split(X_numeric, y_encoded, test_size=0.2, random_state=42)

# Train model
model = LinearRegression()
model.fit(X_train, y_train)

# Predict and evaluate
y_pred = model.predict(X_test)
print("R² score:", r2_score(y_test, y_pred))

# Calculate MSE
mse = mean_squared_error(y_test, y_pred)

# Calculate RMSE by taking the square root of the MSE
print("RMSE:", np.sqrt(mse))

R² score: 0.7442130858002536
RMSE: 0.25285793430805237


You've trained a linear regression model to predict whether a website is phishing or legitimate, based on the numeric features of the website. You've evaluated its performance using two metrics: R-squared (R²) and Root Mean Squared Error (RMSE).

Here's what those values mean:

R² score: 0.7442130858002536

The R-squared value, also known as the coefficient of determination, represents the proportion of the variance in the dependent variable (whether the website is phishing or legitimate) that is predictable from the independent variables (the numeric features).
An R² of 0.744 (or approximately 74.4%) means that your model explains about 74.4% of the variability in the target variable.
In simpler terms, your model does a reasonably good job of capturing the relationship between the website features and its status (phishing or legitimate). A higher R² generally indicates a better fit of the model to the data, though a high R² doesn't necessarily mean the model is perfect or that the variables are the only factors.
RMSE: 0.25285793430805237

RMSE stands for Root Mean Squared Error. It measures the average magnitude of the errors between the predicted values and the actual values.
The unit of RMSE is the same as the unit of the dependent variable. In your case, your dependent variable y_encoded is either 0 (legitimate) or 1 (phishing).
An RMSE of approximately 0.253 means that, on average, your model's predictions are about 0.253 units away from the actual value (0 or 1).
Since your target variable is binary (0 or 1), an RMSE of 0.253 indicates that the model's predictions, which are continuous values between 0 and 1, deviate from the true binary status by this amount on average. A lower RMSE indicates that the model's predictions are closer to the actual values, which is desirable.
In summary:

Your model is able to explain a significant portion (about 74.4%) of the variation in whether a website is phishing or not based on its numeric features.
The average error in your model's predictions, in terms of the 0/1 scale, is about 0.253.
These metrics suggest that your linear regression model has some predictive power for classifying websites as phishing or legitimate based on the numeric features, although it's not a perfect prediction. Since your target variable is binary, it's worth noting that while you've used a linear regression model (which is typically for continuous outcomes), the results are interpreted in the context of predicting a value between 0 and 1. For binary classification problems, metrics like accuracy, precision, recall, F1-score, and ROC AUC are often more appropriate, and models designed for classification (like Logistic Regression, Support Vector Machines, or Tree-based models) are typically used.

In [ ]:
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2, include_bias=False, interaction_only=False)
X_poly = poly.fit_transform(X)

# Train model again
X_train_poly, X_test_poly, y_train, y_test = train_test_split(X_poly, y, test_size=0.2, random_state=42)
model_poly = LinearRegression()
model_poly.fit(X_train_poly, y_train)

# Predict and evaluate
y_pred_poly = model_poly.predict(X_test_poly)
print("R² (Poly):", r2_score(y_test, y_pred_poly))
print("RMSE (Poly):", mean_squared_error(y_test, y_pred_poly, squared=False))


ValueError: could not convert string to float: 'http://www.crestonwood.com/router.php'